cpp/src/io/csv/writer_impl.cu

```diff
@@ -0,0 +1,635 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file writer_impl.cu
+ * @brief cuDF-IO CSV writer class implementation
+ */
+
+#include "writer_impl.hpp"
+
+#include <cudf/copying.hpp>
+#include <cudf/null_mask.hpp>
+
+#include <cudf/utilities/traits.hpp>
+
+#include <cudf/strings/convert/convert_booleans.hpp>
+#include <cudf/strings/convert/convert_integers.hpp>
+#include <cudf/strings/convert/convert_floats.hpp>
+#include <cudf/strings/convert/convert_datetime.hpp>
+
+#include <cudf/strings/replace.hpp>
+#include <cudf/strings/combine.hpp>
+
+#include <strings/utilities.cuh>
+
+#include <algorithm>
+#include <cstring>
+#include <utility>
+#include <type_traits>
+#include <iterator>
+#include <sstream>
+
+#include <thrust/scan.h>
+#include <thrust/execution_policy.h>
+#include <thrust/transform.h>
+#include <thrust/count.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <rmm/device_buffer.hpp>
+
+
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/strings/detail/modify_strings.cuh>
+
+namespace cudf {
+namespace experimental {
+namespace io {
+namespace detail {
+namespace csv {
+
+namespace {//anonym.
+//helpers:
+
+using namespace cudf::strings;
+
+//predicate to determine if given string_view conatins special characters:
+//{"\"", "\n", <delimiter>}
+//
+struct predicate_special_chars
+{
+  explicit predicate_special_chars(string_scalar const& delimiter,
+                                   cudaStream_t stream = 0):
+    delimiter_(delimiter.value(stream))
+  {
+  }
+  
+  __device__
+  bool operator()(string_view const& str_view) const {
+    //if (any_of{"\"", "\n", <delimiter>} )
+    //
+    char const* quote_str = "\"";
+    char const* newline_str = "\n";
+    size_type len1byte{1};//<-TODO: confirm that find() expects len==1, or 2 (including '\0')
+    
+    if( (str_view.find(quote_str, len1byte) >= 0) ||
+        (str_view.find(newline_str, len1byte) >= 0) ||
+        (str_view.find(delimiter_) >= 0) ) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+private:
+  string_view delimiter_;
+};
+
+  
+struct probe_special_chars
+{
+  probe_special_chars(column_device_view const d_column,
+                      predicate_special_chars const& predicate):
+    d_column_(d_column),
+    predicate_(predicate)
+  {
+  }
+
+  __device__
+  int32_t operator()(size_type idx) const {
+    if( d_column_.is_null(idx) )
+      return 0; // null string
+
+    string_view d_str = d_column_.template element<string_view>(idx);
+    
+    if( predicate_(d_str) ) {  
+      //count number of quotes "\""
+      size_type num_quotes{0};
+      char const quote_char = '\"';
+
+      //TODO: confirm below is correct;
```

Comment:
Looks ok to me; what's the issue?

Reply:
The question was if there was a `operator == (char_utf8, char)`. It looks like there is, so not an issue anymore. Will remove the comment.